added pretty-printer binary

Sanket-0510 opened the pull request:
This PR is the demonstration of simple p4c backend with minimal frontend passes, printing the formatted p4c program using the pretty-printer. Use can pass a `.p4` file to this `./p4fmt` binary and it will print the formatted p4 program for you.

to use pretty-printer clone this repo
` git clone github.com/Sanket-0510/p4c`

`cd pretty-printer`

`./p4fmt ./example.p4`

I have added a sample p4 program in example.p4 but you can pass any p4 file to this binary and it will format the code for you.

qobilidop commented:
Hi @Sanket-0510,

Could you add the source code to this PR, and not the compiled binary? You can move it out from the extensions directory so it doesn't [gets ignored](https://github.com/Sanket-0510/p4c/blob/386e56ccef01f8e459479dafcf326e69ee34203f/.gitignore#L14). Also please figure out the necessary cmake changes to enable the compiling.

Thanks!

Sanket-0510 committed:
Configured the backend for formatter example, added verious options for priting at different stages

Sanket-0510 committed:
fix the linking error

Sanket-0510 committed:
added readme for the description

Sanket-0510 commented:
> Hi @Sanket-0510,
> 
> Could you add the source code to this PR, and not the compiled binary? You can move it out from the extensions directory so it doesn't [gets ignored](https://github.com/Sanket-0510/p4c/blob/386e56ccef01f8e459479dafcf326e69ee34203f/.gitignore#L14). Also please figure out the necessary cmake changes to enable the compiling.
> 
> Thanks!

Hey @qobilidop I was stuck into CMake Linking Error hence I had only pushed the binary at first but now you can build binary using this branch. I have also added some useful options which can be used to print formatted output efficiently. 

I also attempted to code `gofmt -w ` type feature which will output the formatted code to the same file but I could not do that. It seems like the code is read in chunks in input so If I try to outStream the formatted code to the same file everything just gets vanished. Maybe some fundamental class changes must required for that feature, neverthless I just attempted it, we can discuss on to that if I get selected. But surely I am really excited for this project.

Regards,
Sanket

Sanket-0510 commented:
> > Hi @Sanket-0510,
> > 
> > Could you add the source code to this PR, and not the compiled binary? You can move it out from the extensions directory so it doesn't [gets ignored](https://github.com/Sanket-0510/p4c/blob/386e56ccef01f8e459479dafcf326e69ee34203f/.gitignore#L14). Also please figure out the necessary cmake changes to enable the compiling.
> > 
> > Thanks!
> 
> Hey @qobilidop I was stuck into CMake Linking Error hence I had only pushed the binary at first but now you can build binary using this branch. I have also added some useful options which can be used to print formatted output efficiently. 
> 
> I also attempted to code `gofmt -w ` type feature which will output the formatted code to the same file but I could not do that. It seems like the code is read in chunks in input so If I try to outStream the formatted code to the same file everything just gets vanished. Maybe some fundamental class changes must required for that feature, neverthless I just attempted it, we can discuss on to that if I get selected. But surely I am really excited for this project.
> 
> Regards,
> Sanket

Hey @qobilidop glad to share that I implemented this option successfully just some twiks and turns with input output files and now we can have formatted output to the same file only. 

I will send you the dummy video as well plus will push code that as well🎉🎉🎉
